### PR TITLE
FormationEnergyDiagram Fix

### PR DESCRIPTION
**Pull request by nwinner**
`core.py`: Minor changes. Some docstrings were wrong. `__repr__()` added to Adsorbate

`thermo.py`: One major fix and new additions

- Fix 1: Adjusted PD entries in the old FormationEnergyDiagram produces incorrect results. This is because the "energy" of the new entry is set by setting the energy with the formation energy. Formation energy *includes* corrections, which means when corrections are added back, there will be double counting. Issue shows up twice in the old code. One instance shown below. Two solutions possible, one is to not include corrections in the derived PD, which loses information, the second is to set energy by subtracting the corrections. The latter is chosen by this patch
- Addition 1: New `MultiFormationEnergyDiagram` for containing many `FormationEnergyDiagram` objects for individual defect types. `MultiFormationEnergyDiagram` can be used to define methods that require all defects, not just one. One example has been included, which is solving for the equilibrium fermi level given a `Dos`.
- Addition 2: First pass at a phase diagram plotter (moved to another PR)

### CODE REVIEW

File: pymatgen/analysis/defects/thermo.py [hunk 1]
```diff
@@ -176,22 +191,25 @@ def __post_init__(self):
         entries = []
         for entry in self.phase_diagram.stable_entries:
             d_ = entry.as_dict()
-            d_["energy"] = self.phase_diagram.get_form_energy(entry)
+            d_["energy"] = self.phase_diagram.get_form_energy(entry) - entry.correction
```
Comment (jmmshn) [inline pymatgen/analysis/defects/thermo.py:194]: So this was initially sloppy on my part. Thanks for catching that!
There another place where the correction was subtracted in the PR, but that was in a dictionary without the correction.
I think we can use a more stripped down dict here without the correction as well.

**Comment by jmmshn** [comment]
OK, so some random change to how setuptools reads pyprojects.toml was breaking tests.
Should be fixed now

**Comment by jmmshn** [comment]
I fixed the tests, but maybe we should coordinate with Matt M on the plotting part since he might have already started.

**Comment by nwinner** [comment]
> I fixed the tests, but maybe we should coordinate with Matt M on the plotting part since he might have already started.

Matt M stopped working on plotting when he found out I was going to do it anyway. Happy to get input from him if you want though.

**Comment by jmmshn** [comment]
@nwinner can you remove the plotting function from this for now and just PR the Mutli-Formation energy diagram part with tests?

**Comment by nwinner** [comment]
Done. I'll make a new PR for it once this is merged

File: pymatgen/analysis/defects/core.py [hunk 2]
```diff
@@ -522,7 +522,11 @@ def __repr__(self) -> str:
 
     def get_multiplicity(self) -> int:
         """Determine the multiplicity of the defect site within the structure."""
-        raise NotImplementedError("Complex defect multiplicity is not implemented.")
+        symm_struct = self.defects[0].symmetrized_structure
+        for defect in self.defects:
+            defect_site = self.structure[defect.defect_site_index]
+            update_structure(symm_struct, defect_site, defect_type=defect.defect_type)
```
Comment (jmmshn) [inline pymatgen/analysis/defects/core.py:528]: So I'm not sure how this should interact with interstitial that's kinda why I did not put this in here.
There is also the problem of how to properly use the symm_structure to get the multiplicity when you are looking at more than one site.  Ideally, you should add a fictitious atom between all the sites to fix the periodic boundary conditions (this is what I used to tag symmetry-distinct hops in the diffusion package).

Comment (jmmshn) [reply pymatgen/analysis/defects/core.py:528]: We might address these better once you have added tests.

Comment (nwinner) [reply pymatgen/analysis/defects/core.py:528]: All it would happen is that interstitials used to build defect complexes would need to have multiplicity set by their generator. 

If you manually create one, it throws the same error that multiplicity cannot be auto-determined.

Comment (jmmshn) [reply pymatgen/analysis/defects/core.py:528]: I don't think that's true because you are returning the `len(symm_struct.find_equivalent_sites(defect_site))` below so you are relying on the symmetry analysis to get the multiplicity for you.

Comment (jmmshn) [reply pymatgen/analysis/defects/core.py:528]: I think for `symm_struct.find_equivalent_sites(defect_site))` to spit out the "right" multiplicity you would have to use a special structure that has a dummy atom inserted already at each symmetry distinct position in the unit-cell right?  Otherwise wouldn't this always just return one for all interstitials?

Comment (nwinner) [reply pymatgen/analysis/defects/core.py:528]: Oh you're right that's my bad. I don't have a solution for that so I'll remove this and punt it to a later date.

Comment (jmmshn) [reply pymatgen/analysis/defects/core.py]: Also, I think this opens up a can of worms regarding how you define a multiplicity of a defect complex.  Like if in interstitial as multiplicity 4 is placed next to substitution with multiplicity 3 what's the right answer?

I think the right answer is to take all possible symmetry operations of the pristine, then apply them to the defect sites then create many of these different defect-inserted structures and see which ones are equivalent.  I've doe this for single sites before but for multiple sites there might be further issues and lots of testing required.

Comment (jmmshn) [reply pymatgen/analysis/defects/core.py]: yeah like it's all doable but it just requires infinite time and grunt work...

Comment (nwinner) [reply pymatgen/analysis/defects/core.py]: A problem for later I suppose.

File: pymatgen/analysis/defects/thermo.py [hunk 3]
```diff
@@ -177,28 +203,44 @@ def __post_init__(self):
...
         if self.inc_inf_values:
             self._chempot_limits_arr = chempot_limits
         else:
             boundary_value = self.chempot_diagram.default_min_limit
             self._chempot_limits_arr = chempot_limits[
                 ~np.any(chempot_limits == boundary_value, axis=1)
             ]
+        self._chempot_limits_arr = self._chempot_limits_arr.dot(
+            1 / self.bulk_entry.composition.reduced_composition.num_atoms
+        )
```
Comment (jmmshn) [inline pymatgen/analysis/defects/thermo.py:243]: OK, this makes sense sorry I missed that!

File: pymatgen/analysis/defects/thermo.py [hunk 4]
```diff
@@ -428,19 +473,20 @@ def get_transitions(
                 potential diagram.
 
         Returns:
-            list[tuple[float, float]]:
-                Transition levels and the formation energy at each transition level.
-                The first and last points are the intercepts with the
-                VBM and CBM respectively.
+            Transition levels and the formation energy at each transition level.
+            Organized as a dictionary keyed by defect __repr__
```
Comment (jmmshn) [inline pymatgen/analysis/defects/thermo.py:477]: This is still the docstring of the single Formation Eneryg

**Review by jmmshn — CHANGES_REQUESTED** [review]
Small change required in docstring
